test/sync/src/commonMain/kotlin/io/realm/entities/sync/SyncObjectWithAllTypes.kt

```diff
@@ -0,0 +1,260 @@
+ */
+
+package io.realm.entities.sync
+
+import io.realm.ObjectId
+import io.realm.RealmInstant
+import io.realm.RealmList
+import io.realm.RealmObject
+import io.realm.annotations.PrimaryKey
+import io.realm.realmListOf
+import io.realm.test.util.TypeDescriptor
+import kotlin.random.Random
+
+private typealias FieldDataFactory = (SyncObjectWithAllTypes) -> Unit
+private typealias FieldValidator = (SyncObjectWithAllTypes) -> Unit
+
+class SyncObjectWithAllTypes : RealmObject {
+    @PrimaryKey
+    var _id: String = "id-${Random.nextLong()}"
+
+    // Non-nullable types
+    var stringField: String = "hello world"
+    var byteField: Byte = 0
+    var charField: Char = 0.toChar()
+    var shortField: Short = 0
+    var intField: Int = 0
+    var longField: Long = 0
+    var booleanField: Boolean = true
+    var doubleField: Double = 0.0
+    var floatField: Float = 0.0.toFloat()
+    var realmInstantField: RealmInstant = RealmInstant.MIN
+    var objectIdField: ObjectId = ObjectId.create()
+    var objectField: SyncObjectWithAllTypes? = null
+
+    // Nullable types
+    var stringNullableField: String? = null
+    var byteNullableField: Byte? = null
+    var charNullableField: Char? = null
+    var shortNullableField: Short? = null
+    var intNullableField: Int? = null
+    var longNullableField: Long? = null
+    var booleanNullableField: Boolean? = null
+    var doubleNullableField: Double? = null
+    var floatNullableField: Float? = null
+    var realmInstantNullableField: RealmInstant? = null
+    var objectIdNullableField: ObjectId? = null
+    var objectNullableField: SyncObjectWithAllTypes? = null
+
+    // RealmLists
+    var stringRealmList: RealmList<String> = realmListOf("hello world")
+    var byteRealmList: RealmList<Byte> = realmListOf(0)
+    var charRealmList: RealmList<Char> = realmListOf(0.toChar())
+    var shortRealmList: RealmList<Short> = realmListOf(0)
+    var intRealmList: RealmList<Int> = realmListOf(0)
+    var longRealmList: RealmList<Long> = realmListOf(0)
+    var booleanRealmList: RealmList<Boolean> = realmListOf(true)
+    var doubleRealmList: RealmList<Double> = realmListOf(0.0)
+    var floatRealmList: RealmList<Float> = realmListOf(0.0.toFloat())
+    var realmInstantRealmList: RealmList<RealmInstant> = realmListOf(RealmInstant.MIN)
+    var objectIdRealmList: RealmList<ObjectId> = realmListOf(ObjectId.create())
+    var objectRealmList: RealmList<SyncObjectWithAllTypes> = realmListOf()
+
+    // Nullable RealmLists of primitive values, not currently supported by Sync
+    // Nullable Object lists, not currently supported by Core
+
+    companion object {
+
+        // Mapping between each Core Field type and functions that can insert data for that type
+        // and also verify the value. This can be used to test objects that has been roundtripped
+        // through Sync.
+        val mapper: Map<TypeDescriptor.CoreFieldType, Pair<FieldDataFactory, FieldValidator>> = createMapper()
```

Comment:
We now have the user facing `RealmStorageType` so if you don't need anything internal things from `CoreFieldType` you could just use that.